File: android-app/src/main/java/fr/cph/chicago/core/listener/OnMarkerClickListener.java
package fr.cph.chicago.core.listener;

import android.support.annotation.NonNull;
import android.util.Log;

import com.annimon.stream.Stream;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.model.Marker;

import fr.cph.chicago.core.fragment.NearbyFragment;
import fr.cph.chicago.data.MarkerDataHolder;
import fr.cph.chicago.entity.AStation;
import fr.cph.chicago.entity.BikeStation;
import fr.cph.chicago.entity.BusStop;
import fr.cph.chicago.entity.Station;
import fr.cph.chicago.entity.dto.BusArrivalRouteDTO;
import fr.cph.chicago.rx.observable.ObservableUtil;

public class OnMarkerClickListener implements GoogleMap.OnMarkerClickListener {

    private static final String TAG = OnMarkerClickListener.class.getSimpleName();

    private final NearbyFragment nearbyFragment;
    private final MarkerDataHolder markerDataHolder;

    public OnMarkerClickListener(final MarkerDataHolder markerDataHolder, final NearbyFragment nearbyFragment) {
        this.markerDataHolder = markerDataHolder;
        this.nearbyFragment = nearbyFragment;
    }

    @Override
    public boolean onMarkerClick(final Marker marker) {
        nearbyFragment.showProgress(true);
        final AStation station = markerDataHolder.getStation(marker);
        if (nearbyFragment.getLayoutContainer().getChildCount() != 0) {
            nearbyFragment.getLayoutContainer().removeViewAt(0);
        }
        loadArrivals(station);
        return false;
    }


    private void loadArrivals(@NonNull final AStation station) {
        if (station instanceof Station) {
            loadTrainArrivals((Station) station);
        } else if (station instanceof BusStop) {
            loadBusArrivals((BusStop) station);
        } else if (station instanceof BikeStation) {
            loadBikes((BikeStation) station);
        }
    }

    private void loadTrainArrivals(final Station trainStation) {
        nearbyFragment.getSlidingUpAdapter().updateTitleTrain(trainStation.getName());
        ObservableUtil.INSTANCE.createTrainArrivalsObservable(nearbyFragment.getContext(), trainStation)
            .subscribe(
                nearbyFragment.getSlidingUpAdapter()::addTrainStation,
                onError -> Log.e(TAG, onError.getMessage(), onError)
            );
    }

    private void loadBusArrivals(final BusStop busStop) {
        nearbyFragment.getSlidingUpAdapter().updateTitleBus(busStop.getName());
        ObservableUtil.INSTANCE.createBusArrivalsObservable(nearbyFragment.getContext(), busStop)
            .subscribe(
                result -> {
                    final BusArrivalRouteDTO busArrivalRouteDTO = new BusArrivalRouteDTO();
                    Stream.of(result).forEach(busArrivalRouteDTO::addBusArrival);
                    nearbyFragment.getSlidingUpAdapter().addBusArrival(busArrivalRouteDTO);
                },
                onError -> Log.e(TAG, onError.getMessage(), onError)
            );
    }

    private void loadBikes(final BikeStation bikeStation) {
        nearbyFragment.getSlidingUpAdapter().updateTitleBike(bikeStation.getName());
        ObservableUtil.INSTANCE.createBikeStationsObservable(bikeStation)
            .subscribe(
                nearbyFragment.getSlidingUpAdapter()::addBike,
                onError -> Log.e(TAG, onError.getMessage(), onError)
            );
    }
}

File: android-app/src/main/kotlin/fr/cph/chicago/core/listener/OnMarkerClickListener.kt
/**
 * Copyright 2018 Carl-Philipp Harmant
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.cph.chicago.core.listener

import android.util.Log
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.Marker
import fr.cph.chicago.core.fragment.NearbyFragment
import fr.cph.chicago.core.model.Station
import fr.cph.chicago.core.model.BikeStation
import fr.cph.chicago.core.model.BusStop
import fr.cph.chicago.core.model.TrainStation
import fr.cph.chicago.core.model.dto.BusArrivalRouteDTO
import fr.cph.chicago.marker.MarkerDataHolder
import fr.cph.chicago.rx.ObservableUtil

class OnMarkerClickListener(private val markerDataHolder: MarkerDataHolder, private val nearbyFragment: NearbyFragment) : GoogleMap.OnMarkerClickListener {

    override fun onMarkerClick(marker: Marker): Boolean {
        nearbyFragment.showProgress(true)
        val station = markerDataHolder.getStation(marker)
        if (nearbyFragment.layoutContainer.childCount != 0) {
            nearbyFragment.layoutContainer.removeViewAt(0)
        }
        //FIXME handle the case a null is returned
        loadArrivals(station!!)
        return false
    }


    private fun loadArrivals(station: Station) {
        when (station) {
            is TrainStation -> loadTrainArrivals(station)
            is BusStop -> loadBusArrivals(station)
            is BikeStation -> loadBikes(station)
        }
    }

    private fun loadTrainArrivals(trainTrainStation: TrainStation) {
        nearbyFragment.slidingUpAdapter.updateTitleTrain(trainTrainStation.name)
        observableUtil.createTrainArrivalsObservable(trainTrainStation)
            .subscribe({ nearbyFragment.slidingUpAdapter.addTrainStation(it) })
            { onError -> Log.e(TAG, onError.message, onError) }
    }

    private fun loadBusArrivals(busStop: BusStop) {
        nearbyFragment.slidingUpAdapter.updateTitleBus(busStop.name)
        observableUtil.createBusArrivalsObservable(busStop)
            .subscribe(
                { result ->
                    val busArrivalRouteDTO = BusArrivalRouteDTO(BusArrivalRouteDTO.busComparator)
                    result.forEach({ busArrivalRouteDTO.addBusArrival(it) })
                    nearbyFragment.slidingUpAdapter.addBusArrival(busArrivalRouteDTO)
                }
            ) { onError -> Log.e(TAG, onError.message, onError) }
    }

    private fun loadBikes(divvyStation: BikeStation) {
        nearbyFragment.slidingUpAdapter.updateTitleBike(divvyStation.name)
        observableUtil.createBikeStationsObservable(divvyStation)
            .subscribe({ nearbyFragment.slidingUpAdapter.addBike(it) })
            { onError -> Log.e(TAG, onError.message, onError) }
    }

    companion object {
        private val TAG = OnMarkerClickListener::class.java.simpleName
        private val observableUtil = ObservableUtil
    }
}
